### PR TITLE
Fix bug in graceful shutdown

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Oops, still had bugs in the last PR. Follow up to #2797

### CODE REVIEW

File: test/Kestrel.Core.Tests/Http2StreamTests.cs [hunk 1]
```diff
@@ -964,12 +964,6 @@ public async Task ContentLength_Received_SingleDataFrameOverSize_Reset()
         [Fact]
         public async Task ContentLength_Received_SingleDataFrameUnderSize_Reset()
         {
-            // I hate doing this, but it avoids exceptions from MemoryPool.Dipose() in debug mode. The problem is since
-            // the stream's ProcessRequestsAsync loop is never awaited by the connection, it's not really possible to
-            // observe when all the blocks are returned. This can be removed after we implement graceful shutdown.
-            Dispose();
-            InitializeConnectionFields(new DiagnosticMemoryPool(KestrelMemoryPool.CreateSlabMemoryPool(), allowLateReturn: true));
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2StreamTests.cs]: This is no longer called in any of our unit tests, right?

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 2]
```diff
@@ -910,8 +910,12 @@ void IHttp2StreamLifetimeHandler.OnStreamCompleted(int streamId)
                         Input.CancelPendingRead();
                     }
 
-                    // Complete the task waiting on all streams to finish
-                    _streamsCompleted.TrySetResult(null);
+
+                    if (_state != Http2ConnectionState.Open)
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Nice catch @Tratcher. Let's add a new regression unit test that:

1. Opens a stream where the application doesn't block.
2. Closes the stream.
3. Opens a new stream where the application blocks.
4. Closes the input pipe
5. Verifies that Http2Connection.ProcessRequestsAsync doesn't complete until the application for the second stream completes.

File: test/Kestrel.Core.Tests/Http2StreamTests.cs (hunk 1 above)
Comment (JunTaoLuo) [reply test/Kestrel.Core.Tests/Http2StreamTests.cs]: There are a few spots:
https://github.com/aspnet/KestrelHttpServer/blob/release/2.2/test/Kestrel.Core.Tests/DiagnosticMemoryPoolTests.cs#L176-L200. This is specifically testing those behaviours so we'll keep these.
https://github.com/aspnet/KestrelHttpServer/blob/release/2.2/test/Kestrel.Transport.FunctionalTests/Http2/ShutdownTests.cs#L100. This is the abortive test and I think it's unavoidable.
https://github.com/aspnet/KestrelHttpServer/blob/0e99235d59031106bfe6c593757b99ff99f261f7/test/Kestrel.Transport.FunctionalTests/MaxRequestBufferSizeTests.cs#L117. I'll see if these can be removed though.

Comment (halter73) [reply test/Kestrel.Core.Tests/Http2StreamTests.cs]: Don't worry about the functional tests. I was only asking about unit tests.

The MaxRequestBufferSizeTests have been that way in a while. Even if it passes without allowing late returns on your machine, I would be very concerned about introducing test flakiness. You're also right about it being unavoidable for the abortive Shutdown tests.

**Comment by JunTaoLuo** [comment]
Added a test

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 3]
```diff
@@ -3324,6 +3324,54 @@ public async Task CompletionSendsFinalGOAWAY()
             VerifyGoAway(await ReceiveFrameAsync(), 0, Http2ErrorCode.NO_ERROR);
         }
 
+        [Fact]
+        public async Task StopProcessingNextRequestSendsGracefulGOAWAYAndWaitsForStreamsToComplete()
+        {
+            var requestSemaphore = new SemaphoreSlim(1);
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: I think this could just be a TCS. No need to block a thread.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 4]
```diff
@@ -3324,6 +3324,54 @@ public async Task CompletionSendsFinalGOAWAY()
...
+            _pair.Application.Input.CancelPendingRead();
+            var result = await readTask;
+            Assert.False(result.IsCompleted);
+
+            // Unblock the request and ProcessRequestsAsync
+            requestSemaphore.Release();
+            await _connectionTask;
+
+            // Assert connection's Input pipe is completed
+            result = await _pair.Application.Input.ReadAsync();
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Nit: .DefaultTimeout()

Comment (halter73) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Also, for consistency, use the same pattern you used above of calling ReadAsync, then CancelPendingRead, then awaiting the task. It should return a completed task before CancelPendingRead is even called if you use a TCS without the RunContinuationsAsync option. This is the one of the few times I would recommend not using that option as long as you add a comment explaining why.

Comment (JunTaoLuo) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: You can't use .DefaultTimeout() here since ReadAsync() returns a ValueTask

Comment (halter73) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: You could always call AsTask() first, but there shouldn't be much risk of a hanging test even if this regresses as long as the extra CancelPendingRead() is added anyway.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 5]
```diff
@@ -3327,11 +3327,11 @@ public async Task CompletionSendsFinalGOAWAY()
...
         {
-            var requestSemaphore = new SemaphoreSlim(1);
-            await InitializeConnectionAsync(context =>
+            var task = Task.CompletedTask;
+            await InitializeConnectionAsync(async context =>
             {
-                requestSemaphore.Wait();
-                return Task.CompletedTask;
+                await task;
+                context.Response.StatusCode = 200;
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Why is this here?

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 6]
```diff
@@ -3347,6 +3347,8 @@ public async Task StopProcessingNextRequestSendsGracefulGOAWAYAndWaitsForStreams
                 withStreamId: 1);
 
             // Send a blocked request
+            var tcs = new TaskCompletionSource<object>(); // Not running continuations as async since the task may be completed
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Nit: Comment on own line. I would phrase it like so:

```
// No RunContinuationsAsynchronously so we can verify assertions immediately after calling SetResult.
```

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs (hunk 5 above)
Comment (JunTaoLuo) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: oops test code I accidentally staged

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 7]
```diff
@@ -3324,6 +3324,54 @@ public async Task CompletionSendsFinalGOAWAY()
...
+            // Assert connection shutdown is still blocked
+            // ProcessRequestsAsync completes the connection's Input pipe
+            var readTask = _pair.Application.Input.ReadAsync();
+            _pair.Application.Input.CancelPendingRead();
+            var result = await readTask;
+            Assert.False(result.IsCompleted);
+
+            // Unblock the request and ProcessRequestsAsync
+            requestSemaphore.Release();
+            await _connectionTask;
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Before you merge, add .DefaultTimeout() here though.

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 8]
```diff
@@ -3324,6 +3324,55 @@ public async Task CompletionSendsFinalGOAWAY()
...
+            // Assert connection shutdown is still blocked
+            // ProcessRequestsAsync completes the connection's Input pipe
+            var readTask = _pair.Application.Input.ReadAsync();
+            _pair.Application.Input.CancelPendingRead();
+            var result = await readTask;
+            Assert.False(result.IsCompleted);
+
+            // Unblock the request and ProcessRequestsAsync
+            tcs.TrySetResult(null);
+            await _connectionTask;
```
Comment (halter73) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Before you merge, add .DefaultTimeout() here though.
